### PR TITLE
Smartcard for stm32h7 uart

### PR DESCRIPTION
**Pull request by ghost**
## Summary
This pull request is intended to show my progress into developing smartcard support for stm32h7.
It is an experimental draft for comment and review.
It shall not be merged yet.

## Testing
A test app is required to exercise the new ioctls.

### CODE REVIEW

**Comment by davids5** [comment]
@slorquet 
A few questions:

Is this for ISO7816? 
Would it be better to make is a separate  driver?

File: include/nuttx/serial/tioctl.h [hunk 1]
```diff
@@ -202,6 +202,23 @@
...
+  {
+    uint8_t enabled : 1; /* Enable smart card mode */
+    uint8_t clken   : 1; /* Enable smart card clock */
+    uint8_t inverse : 1; /* Use inverse convention */
+    uint32_t clkout;     /* Clock output frequency requested,
+                          * updated with actual value, which is the largest
+                          * value inferior to the request */
+    uint16_t etu;        /* Elementary time unit, eg baud rate in clock counts */
+    uint8_t  cgt;        /* Character guard time in ETU units */
+  };
```
Comment (acassis) [inline include/nuttx/serial/tioctl.h:220]: Please align { and } to start of the line, see "struct winsize" below for reference!

File: arch/arm/src/stm32h7/stm32_serial.c [hunk 2]
```diff
@@ -2660,6 +2719,201 @@ static int up_ioctl(struct file *filep, int cmd, unsigned long arg)
...
+        else
+          {
+            /* User has requested to disable smart card mode */
+
+            cr3 &= ~USART_CR3_SCEN;
+            cr2 &= ~USART_CR2_CLKEN; /* Disable smart card clock */
+
+            /* Re-enable RX and disable CLK */
+
+            stm32_configgpio  (priv->rx_gpio);
```
Comment (pkarashchenko) [inline arch/arm/src/stm32h7/stm32_serial.c:2899]:
```suggestion
            stm32_configgpio(priv->rx_gpio);
```

File: arch/arm/src/stm32h7/stm32_serial.c [hunk 3]
```diff
@@ -2660,6 +2719,201 @@ static int up_ioctl(struct file *filep, int cmd, unsigned long arg)
...
+              {
+                /* Compute prescaler to match request,
+                 * then return the actual value.
+                 */
+
+                uint8_t clkdiv = 1;
+                uint32_t divided = priv->apbclock >> 1;
+                while ((divided > priv->smartcard.clkout) && (clkdiv < 31))
+                  {
+                    clkdiv   += 1;
```
Comment (pkarashchenko) [inline arch/arm/src/stm32h7/stm32_serial.c:2849]:
```suggestion
                    clkdiv++;
```

File: include/nuttx/serial/tioctl.h (hunk 1 above)
Comment (slorquet) [reply include/nuttx/serial/tioctl.h:220]: done, thanks

File: arch/arm/src/stm32h7/stm32_serial.c [hunk 4]
```diff
@@ -2660,6 +2719,201 @@ static int up_ioctl(struct file *filep, int cmd, unsigned long arg)
...
+            /* Re-enable RX and disable CLK */
+
+            stm32_configgpio  (priv->rx_gpio);
+            stm32_unconfiggpio(priv->ck_gpio);
+          }
+
+        /* Apply new settings */
+
+        up_serialout(priv, STM32_USART_CR3_OFFSET , cr2);
+        up_serialout(priv, STM32_USART_CR3_OFFSET , cr3);
```
[the remark this answers is not in the document]
Comment (slorquet) [reply arch/arm/src/stm32h7/stm32_serial.c:2906]: The goal was to ensure alignment of the commas. Is that kind of alignment not important for readability? Checkpatch did not notify me about this

File: arch/arm/src/stm32h7/stm32_serial.c (hunk 2 above)
Comment (slorquet) [reply arch/arm/src/stm32h7/stm32_serial.c:2899]: Please tell me globally if that kind of alignment within consecutive lines is a thing I should continue or not.

I personally find it has cosmetic value and has no drawbacks, but I can avoid it.

File: arch/arm/src/stm32h7/stm32_serial.c [hunk 5]
```diff
@@ -2660,6 +2719,201 @@ static int up_ioctl(struct file *filep, int cmd, unsigned long arg)
...
+                /* The refernce clock for the ETU is the divided clock */
+
+                etuclock = priv->apbclock / clkdiv;
+
+                /* Compute baud rate from clock divider and ETU */
+
+                priv->baud      = etuclock / priv->smartcard.etu;
+                priv->bits      = 8;
+                priv->parity    = 2; /* even */
+                priv->stopbits2 = TRUE;
```
[the remark this answers is not in the document]
Comment (slorquet) [reply arch/arm/src/stm32h7/stm32_serial.c:2880]: done

File: arch/arm/src/stm32h7/stm32_serial.c (hunk 3 above)
Comment (slorquet) [reply arch/arm/src/stm32h7/stm32_serial.c:2849]: done.

**Comment by slorquet** [comment]
Thanks @pkarashchenko for the extensive proofreading. I'm waiting more confirmation for the alignments.

@davids5 : Yes, it implements ISO7816, but the framework this goes into is PC/SC, not just ISO7816.
And no, it should not be a separate driver at this point.

This is merely ISO7816 support for the stm32 uart, just like the uart also has LIN SPI or RS485 modes. This is extremely hardware specific, and its just a low level driver.

I want to make it look like PC/SC since this is the de-facto standard for smart cards in C, already implemented in Windows and Linux. It has real value to help people port smart card programs.

In PC/SC, you can have many types of drivers. This is usually handled by the CCID driver, which describes a way to access USB readers. But CCID also support serial attached readers, and proprietary drivers could be implemented in any way possible.

These drivers are then managed by userspace PC/SC daemon that handles access to drivers from application programs.

so to sum up, smart cards are not accessed as devices from user space, but only through a management daemon that hides the details of actual reader access and implementation.

NuttX already has support for usb host, so it could support USB smart card readers via ccid. Thats why I'm keeping this architecture, so USB readers can be handled by the PC/SC daemon in the future. This means the stm 32 uart side of things does not need to have an easy to use user interface.

It's a long text so I can add precisions and details to make it more clear.

**Comment by acassis** [comment]
> @slorquet A few questions:
> 
> Is this for ISO7816? Would it be better to make is a separate driver?

@slorquet I think @davids5 suggestion makes sense!

The 1wire implementation for STM32 is also uses the UART peripheral, but it is a separated file:
arch/arm/src/stm32/stm32_1wire.c

So, this way you could create a drivers/smartcard/iso7816.c that uses this driver as a lowerhalf

**Comment by slorquet** [comment]
Yes, I am worrying that it will be a lot more validation work.

These days I'm back to this development. I still don't get clock output from the clock pin.

(I am using a small trivial test app to exercise this driver)

**Comment by acassis** [comment]
@davids5 @pkarashchenko do you think this PR should be integrated? Seems like Sebastien deleted this Github user and the "ghost" closed it for us
